Add HAL for HC32

shadow578 opened the pull request:
<!--

Submitting a Pull Request

- Please fill out all sections of this form. You can delete the helpful comments.
- Pull Requests without clear information will take longer and may even be rejected.
- We get a high volume of submissions so please be patient during review.

-->

### Description

<!--

Clearly describe the submitted changes with lots of details. Include images where helpful. Initial reviewers may not be familiar with the subject, so be as thorough as possible. You can use MarkDown syntax to improve readability with bullet lists, code blocks, and so on. PREVIEW and fix up formatting before submitting.

-->

this PR adds a HAL for the [HC32F460 series](https://www.hdsc.com.cn/Category83-1487) and compatible MCUs.   

the HAL provides full support for "normal" 3D-printing operation, such as stepper control, heaters, temperature sensors, endstop interrupts, etc..
support for DWIN lcd, BLTouch, and postmortem debugging is confirmed to be working as well.   

currently, the only board i've tested with is the Voxelab Aquila V1.0.1, which is found in the Voxelab Aquila X2. Other HC32-based boards should work too.


for some extra details, please refer to the [README of the HAL](https://github.com/shadow578/Marlin-H32/blob/aquila/Marlin/src/HAL/HC32/README.md).



### Requirements

<!-- Does this PR require a specific board, LCD, etc.? -->

the PR requires a HC32-based board for testing.


### Benefits

<!-- What does this PR fix or improve? -->

by merging this PR, Marlin would support HC32-based printers, allowing more people to replace the potentially dangerous firmware of their printers with Marlin.

Additionally, having a HC32-HAL in the official repo might cause more developers to work on the HC32, improving marlin support over time.



### Configurations

<!-- Attach Configurations ZIP and any other files needed to test this PR. -->

[sample configuration for Aquila X2 w/ DWIN MarlinUI + BLTouch](https://github.com/shadow578/Marlin-H32/files/13325653/Aquila-X2_config.zip)


### Related Issues

<!-- Does this PR fix a bug or fulfill a Feature Request? Link related Issues here. -->

https://github.com/MarlinFirmware/Marlin/issues/24338


### Extra Notes

the commit history is kind of messy since i've been working on this since the start of 2023. 
If possible, this PR should be squash-merged...

EvilGremlin commented:
Yeah, this is too much.
Preferably, make a copy branch for yourself and squash this one.

ellensp commented on Marlin/Configuration.h:
You lost the final new line

shadow578 replied:
oops, for some reason the PR shows some file changes that didn't show for me locally. I've reviewed those unintended changes and (hopefully) reverted them

ellensp commented on Marlin/src/HAL/LPC1768/upload_extra_script.py:
You have removed execute permission of this file, this will break uploading on LPC1768/9

shadow578 replied:
i'm on windows, so idk how i'd do that. I've reverted it to the version on the bugfix-2.1.x branch, so it should be fine.

ellensp commented on Marlin/Configuration_adv.h:
You lost the final new line

shadow578 commented:
> Yeah, this is too much.
> Preferably, make a copy branch for yourself and squash this one.

could you elaborate how i'd do that? 
I thought that the PR could just be squash-merged. Is that not the case? 
Or is there an disadvantage to doing that?

EvilGremlin commented:
https://docs.github.com/ru/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squashing-and-merging-a-long-running-branch
basically, it forces you to force-push

maikramer commented:
I would benefit from this also, because this will add support for HC32F46x based printers. I used his port to port my anycubic cobra max. Then with this merged, i can start using marlin mainline. Thanks @shadow578 .
My port is this one https://github.com/maikramer/MarlinKobraMax

shadow578 commented:
> https://docs.github.com/ru/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squashing-and-merging-a-long-running-branch
basically, it forces you to force-push

ok, how would i go about squashing this then? 
Would this work:
1. pull latest commits from `bugfix-2.1.x`, then create a new branch from that
2. create a PR into the branch created in step 1
3. squash merge the PR
4. create a new PR agains Marlin, merging the branch from step 1
5. close this PR

or is there a better way?

shadow578 committed:
Add HC32 HAL

squashed 161 commits from https://github.com/shadow578/Marlin-H32/tree/aquila

EvilGremlin commented:
Described here https://marlinfw.org/docs/development/getting_started_pull_requests.html

shadow578 commented:
> Described here https://marlinfw.org/docs/development/getting_started_pull_requests.html
Though i prefer git reset --hard way

hope its better now

thinkyhead committed:
headers, formatting

thinkyhead committed:
add'l formatting

thinkyhead committed:
🎨 Update file headers

Marlin is collectively © by the Organization, with individual contributors having © on their original work as documented in the commit history.

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26414

thinkyhead committed:
more formatting

thinkyhead committed:
…and pins

thinkyhead committed:
pins, pins debug

thinkyhead committed:
es ints

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26414

thinkyhead committed:
env adjust, add test

thinkyhead committed:
pins

thinkyhead committed:
Try Python 3.9

thinkyhead committed:
✅ Use Python 3.9 for CI

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26414

thinkyhead committed:
uc comments, format

thinkyhead committed:
etc.

Ivoz commented:
Many thanks for your work on this @shadow578 !! 🍾🍾🥂🎉

Ivoz commented:
@classicrocker883 you know if vanilla build of marlin with this still crashes with an SD card?
